src/metric.rs

```diff
@@ -0,0 +1,305 @@
+use byteorder::WriteBytesExt;
+use memmap::{Mmap, MmapViewSync, Protection};
+use std::ffi::CString;
+use std::io;
+use std::io::Cursor;
+use std::mem;
+
+const ITEM_BIT_LEN: usize = 10;
+
+pub const I32_METRIC_TYPE_CODE: u32 = 0;
+pub const U32_METRIC_TYPE_CODE: u32 = 1;
+pub const I64_METRIC_TYPE_CODE: u32 = 2;
+pub const U64_METRIC_TYPE_CODE: u32 = 3;
+pub const F32_METRIC_TYPE_CODE: u32 = 4;
+pub const F64_METRIC_TYPE_CODE: u32 = 5;
+pub const STRING_METRIC_TYPE_CODE: u32 = 6;
+
+pub trait MetricType {
+    fn type_code(&self) -> u32;
+    fn write_to_writer<W: WriteBytesExt>(&self, writer: &mut W) -> io::Result<()>;
+}
+
+macro_rules! impl_metric_type_for (
+    ($typ:tt, $base_typ:tt, $type_code:expr) => (
+        impl MetricType for $typ {
+            
+            fn type_code(&self) -> u32 {
+                $type_code
+            }
+
+            fn write_to_writer<W: WriteBytesExt>(&self, mut w: &mut W) -> io::Result<()> {
+                w.write_u64::<super::Endian>(
+                    unsafe {
+                        mem::transmute::<$typ, $base_typ>(*self) as u64
+                    }
+                )
+            }
+
+        }
+    )
+);
+
+impl_metric_type_for!(i32, u32, I32_METRIC_TYPE_CODE);
+impl_metric_type_for!(u32, u32, U32_METRIC_TYPE_CODE);
+impl_metric_type_for!(i64, u64, I64_METRIC_TYPE_CODE);
+impl_metric_type_for!(u64, u64, U64_METRIC_TYPE_CODE);
+impl_metric_type_for!(f32, u32, F32_METRIC_TYPE_CODE);
+impl_metric_type_for!(f64, u64, F64_METRIC_TYPE_CODE);
```

Comment:
Actually I think I am wrong here, as my understanding is coming from using literals to represent types (like in speed) instead of using types themselves.